src/Utilities/Compiler/WellKnownTypeProvider.cs

```diff
@@ -46,7 +48,47 @@ public bool TryGetOrCreateTypeByMetadataName(string fullTypeName, out INamedType
         {
             namedTypeSymbol = _fullNameToTypeMap.GetOrAdd(
                 fullTypeName,
-                (string s) => Compilation.GetTypeByMetadataName(s));    // Caching null results in our cache is intended.
+                fullyQualifiedMetadataName =>
+                {
+                    // Caching null results in our cache is intended.
+                    var type = Compilation.GetTypeByMetadataName(fullyQualifiedMetadataName)
+                        ?? Compilation.Assembly.GetTypeByMetadataName(fullyQualifiedMetadataName);
+                    if (type is null)
+                    {
+                        foreach (var module in Compilation.Assembly.Modules)
+                        {
+                            foreach (var referencedAssembly in module.ReferencedAssemblySymbols)
+                            {
+                                var currentType = referencedAssembly.GetTypeByMetadataName(fullyQualifiedMetadataName);
```

Comment:
Observation (not a change request): Any API based analyzer that will bail out for a compilation due to the API not being present in the compilation with end up executing `GetTypeByMetadataName` for each module's assembly references. Hopefully, the API invocation is cheap when the assembly has no such API. If not, we may have to tune this in future.